src/Frontend/Page/Idea.hs

```diff
@@ -88,70 +89,125 @@ import qualified Types (MoveIdea)
 -- * 5.4 Idea detail page: Voting phase
 -- * 5.6 Idea detail page: Feasible / not feasible
 -- * 5.7 Idea detail page: Winner
-data ViewIdea = ViewIdea RenderContext IdeaStats
-  deriving (Eq, Show, Read)
+data ViewIdea = ViewIdea { _viCtx :: CapCtx, _viStats :: IdeaStats }
+  deriving (Eq, Show, Read, Generic)
+
+instance SOP.Generic ViewIdea
+
+makeLenses ''ViewIdea
 
 instance Page ViewIdea where
+    isAuthorized = authNeedCaps [CanView] viCtx
 
 data ViewDeletedIdea = ViewDeletedIdea Idea
   deriving (Eq, Show, Read)
 
-instance Page ViewDeletedIdea where
-
 -- | 6. Create idea
-data CreateIdea = CreateIdea IdeaLocation
-  deriving (Eq, Show, Read)
+data CreateIdea = CreateIdea { _ciCtx :: CapCtx, _ciLoc :: IdeaLocation }
+  deriving (Eq, Show, Read, Generic)
+
+instance SOP.Generic CreateIdea
+
+makeLenses ''CreateIdea
 
 instance Page CreateIdea where
+    isAuthorized = authNeedCaps [CanCreateIdea] ciCtx
 
 -- | 7. Edit idea
-data EditIdea = EditIdea Idea
-  deriving (Eq, Show, Read)
+data EditIdea = EditIdea { _eiCtx :: CapCtx, _eiIdea :: Idea }
+  deriving (Eq, Show, Read, Generic)
+
+instance SOP.Generic EditIdea
+
+makeLenses ''EditIdea
 
 instance Page EditIdea where
+    isAuthorized = authNeedCaps [CanEditAndDelete] eiCtx
 
 -- | X. Move idea
 -- Move idea to a topic.
-data MoveIdea = MoveIdea Idea [Topic]
-  deriving (Eq, Show, Read)
+data MoveIdea = MoveIdea { _miCtx :: CapCtx, _miIdea :: Idea, _miTopicChoices :: [Topic] }
+  deriving (Eq, Show, Read, Generic)
+
+instance SOP.Generic MoveIdea
+
+makeLenses ''MoveIdea
 
 instance Page MoveIdea where
+    isAuthorized = authNeedCaps [CanMoveBetweenLocations] miCtx
 
-data ReportIdea = ReportIdea Idea
-  deriving (Eq, Show, Read)
+data ReportIdea = ReportIdea { _riCtx :: CapCtx, _riIdea :: Idea }
+  deriving (Eq, Show, Read, Generic)
+
+instance SOP.Generic ReportIdea
+
+makeLenses ''ReportIdea
 
 instance Page ReportIdea where
+    -- You can report as soon as you can view the idea.
+    isAuthorized = authNeedCaps [CanView] riCtx
 
 -- | X. Comment idea
-data CommentOnIdea = CommentOnIdea Idea (Maybe Comment)
-  deriving (Eq, Show, Read)
+data CommentOnIdea = CommentOnIdea
+    { _coiCtx :: CapCtx, _coiIdea :: Idea, _coiComment :: Maybe Comment }
+  deriving (Eq, Show, Read, Generic)
+
+instance SOP.Generic CommentOnIdea
+
+makeLenses ''CommentOnIdea
 
 instance Page CommentOnIdea where
+    isAuthorized = authNeedCaps [CanComment] coiCtx
 
 -- | X. Deem idea feasible / not feasible
 -- Assumption: The idea is located in the topic (via 'IdeaLocation').
-data JudgeIdea = JudgeIdea IdeaJuryResultType Idea Topic
-  deriving (Eq, Show, Read)
+data JudgeIdea = JudgeIdea
+    { _jiCtx    :: CapCtx
+    , _jiResult :: IdeaJuryResultType
+    , _jiIdea   :: Idea
+    , _jiTopic  :: Topic
+    }
+  deriving (Eq, Show, Read, Generic)
+
+instance SOP.Generic JudgeIdea
+
+makeLenses ''JudgeIdea
 
 instance Page JudgeIdea where
+    isAuthorized = authNeedCaps [CanMarkFeasiblity] jiCtx
```

Comment:
rename cap to `CanJudge`.  (not necessarily in this PR, but since it's an easy change...)